File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.py
import heapq

def s2t(s):
    h, m = map(int, s.split(":"))
    return h * 60 + m

def solution(book_time):
    # 1. 시각 정수화 + 정렬
    time = [(s2t(st), s2t(et)) for st, et in book_time]
    time.sort(key=lambda x: (x[0], x[1]))  # 입실 빠른 순

    q = []  # 종료시간 + 청소시간 min-heap (우선순위 큐)

    for st, et in time:
        # 2. 가장 빨리 끝나는 방을 하나만 꺼내서 확인
        if q and q[0] <= st:
            heapq.heappop(q)  # 방 재사용 가능

        # 3. 새 종료시간 + 청소시간 push
        heapq.heappush(q, et + 10)

    # 4. heap 크기 = 필요한 방의 수
    return len(q)

File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.java
import java.util.*;
class Solution {
    public int solution(String[][] book_time) {
        int ans = 0;
        int n = book_time.length;
        int[][] arr = new int[n][2];
        for (int i = 0; i < n; i++) {
            String[] bt = book_time[i];
            arr[i][0] = s2t(bt[0]);
            arr[i][1] = s2t(bt[1]) + 10;
        }
        // 일찍 시작 기준으로 정렬
        Arrays.sort(arr,Comparator
                    .comparingInt((int[] a) -> a[0])
                   .thenComparingInt((int[] a) -> a[1]));
        
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        for (int[] a : arr) {
            int st = a[0];
            int et = a[1];
            while(!pq.isEmpty() && pq.peek() <= st) {
                pq.poll();
            }
            pq.add(et);
            ans = Math.max(ans, pq.size());
        }
        return ans;
    }
    
    public int s2t(String str) {
        String[] s = str.split(":");
        return Integer.valueOf(s[0])*60 + Integer.valueOf(s[1]);
    }
}